File: camera_helper/src/org/lineageos/camerahelper/KeyHandler.java
/*
 * Copyright (C) 2019 The LineageOS Project
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.camerahelper;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.KeyEvent;
import android.view.WindowManager;

import com.android.internal.os.DeviceKeyHandler;

public class KeyHandler implements DeviceKeyHandler {
    private static final String TAG = KeyHandler.class.getSimpleName();

    // Camera motor event key codes
    private static final int MOTOR_EVENT_MANUAL_TO_DOWN = 184;
    private static final int MOTOR_EVENT_UP_ABNORMAL = 186;
    private static final int MOTOR_EVENT_DOWN_ABNORMAL = 189;

    private final Context mContext;

    public KeyHandler(Context context) {
        mContext = context;
    }

    public KeyEvent handleKeyEvent(KeyEvent event) {
        int scanCode = event.getScanCode();

        switch (scanCode) {
            case MOTOR_EVENT_MANUAL_TO_DOWN:
            case MOTOR_EVENT_UP_ABNORMAL:
            case MOTOR_EVENT_DOWN_ABNORMAL:
                if (event.getAction() == KeyEvent.ACTION_DOWN) {
                    if (scanCode == MOTOR_EVENT_MANUAL_TO_DOWN) {
                        closeAllApps();
                    }
                    new Handler(Looper.getMainLooper()).post(() -> {
                        AlertDialog alertDialog = showCameraMotorWarning(scanCode);
                        if (alertDialog != null) {
                            alertDialog.getWindow().setType(WindowManager.LayoutParams.TYPE_SYSTEM_ALERT);
                            alertDialog.setCanceledOnTouchOutside(false);
                            alertDialog.show();
                        }
                    });
                }
                return null;
            default:
                return event;
        }
    }

    private void closeAllApps() {
        // Go back to home to close all camera apps first
        Intent intent = new Intent(Intent.ACTION_MAIN);
        intent.addCategory(Intent.CATEGORY_HOME);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        mContext.startActivity(intent);
    }

    private AlertDialog showCameraMotorWarning(int scanCode) {
        Context packageContext = null;
        try {
            packageContext = mContext.createPackageContext("org.lineageos.camerahelper", 0);
        } catch (NameNotFoundException | SecurityException e) {
            Log.e(TAG, "Failed to create package context", e);
        }

        if (packageContext != null)
            return null;

        AlertDialog.Builder builder = new AlertDialog.Builder(packageContext);
        switch (scanCode) {
            case MOTOR_EVENT_MANUAL_TO_DOWN:
                builder.setTitle(R.string.warning);
                builder.setMessage(R.string.motor_press_message);
                builder.setPositiveButton(android.R.string.ok, null);
                break;
            case MOTOR_EVENT_UP_ABNORMAL:
                builder.setTitle(R.string.warning);
                builder.setMessage(R.string.motor_cannot_go_up_message);
                builder.setNegativeButton(R.string.retry, (dialog, which) -> {
                    CameraMotorController.setOpen(true);
                });
                builder.setPositiveButton(R.string.close, (dialog, which) -> {
                    CameraMotorController.setOpen(false);
                    closeAllApps();
                });
                break;
            case MOTOR_EVENT_DOWN_ABNORMAL:
                builder.setTitle(R.string.warning);
                builder.setMessage(R.string.motor_cannot_go_down_message);
                builder.setPositiveButton(R.string.retry, (dialog, which) -> {
                    // Close the camera
                    CameraMotorController.setOpen(false);
                });
                break;
            default:
                return null;
        }
        return builder.create();
    }
}

File: camera_helper/src/org/lineageos/camerahelper/KeyHandler.kt
/*
 * Copyright (C) 2019 The LineageOS Project
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package org.lineageos.camerahelper

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.view.KeyEvent
import android.view.WindowManager
import com.android.internal.os.DeviceKeyHandler

class KeyHandler(private val context: Context) : DeviceKeyHandler {

    override fun handleKeyEvent(event: KeyEvent): KeyEvent? {
        when (event.scanCode) {
            MOTOR_EVENT_MANUAL_TO_DOWN,
            MOTOR_EVENT_UP_ABNORMAL,
            MOTOR_EVENT_DOWN_ABNORMAL -> {
                if (event.action != KeyEvent.ACTION_DOWN) {
                    return null
                }

                if (event.scanCode == MOTOR_EVENT_MANUAL_TO_DOWN) {
                    closeAllApps()
                }

                Handler(Looper.getMainLooper()).post {
                    showCameraMotorWarning(event.scanCode)?.apply {
                        window?.setType(WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY)
                        setCanceledOnTouchOutside(false)
                        show()
                    }
                }
                return null
            }
            else -> return event
        }
    }

    private fun closeAllApps() {
        // Go back to home to close all camera apps first
        val intent = Intent(Intent.ACTION_MAIN).apply {
            addCategory(Intent.CATEGORY_HOME)
            setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)
    }

    private fun showCameraMotorWarning(scanCode: Int): AlertDialog? {
        val packageContext = runCatching {
            context.createPackageContext(
                KeyHandler::class.java.`package`!!.name,
                0
            )
        }.getOrElse { return null }

        return when (scanCode) {
            MOTOR_EVENT_MANUAL_TO_DOWN -> {
                AlertDialog.Builder(packageContext).apply {
                    setTitle(R.string.warning)
                    setMessage(R.string.motor_press_message)
                    setPositiveButton(android.R.string.ok, null)
                }.create()
            }
            MOTOR_EVENT_UP_ABNORMAL -> {
                AlertDialog.Builder(packageContext).apply {
                    setTitle(R.string.warning)
                    setMessage(R.string.motor_cannot_go_up_message)
                    setNegativeButton(R.string.retry) { _, _ ->
                        CameraMotorController.setMotorDirection("1")
                    }
                    setPositiveButton(R.string.close) { _, _ ->
                        CameraMotorController.setMotorDirection("0")
                        closeAllApps()
                    }
                }.create()
            }
            MOTOR_EVENT_DOWN_ABNORMAL -> {
                AlertDialog.Builder(packageContext).apply {
                    setMessage(R.string.motor_cannot_go_down_message)
                    setPositiveButton(R.string.retry) { _, _ ->
                        CameraMotorController.setMotorDirection("0")
                   }
                }.create()
                }
            else -> null
        }
    }


    companion object {
        // Camera motor event key codes
        private const val MOTOR_EVENT_MANUAL_TO_DOWN = 184
        private const val MOTOR_EVENT_UP_ABNORMAL = 186
        private const val MOTOR_EVENT_DOWN_ABNORMAL = 189
    }
}
